include/eve/algo/container/detail/soa_storage2.hpp

```diff
@@ -0,0 +1,172 @@
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data_aligned() const
+    {
+      auto ptrs = kumi::map([](auto const* m) { return as_aligned_pointer(m); }, indexes_);
+      return kumi::apply(eve::algo::views::zip, ptrs);
+    }
+
+    EVE_FORCEINLINE auto data()       { return eve::unalign(data_aligned()); }
+    EVE_FORCEINLINE auto data() const { return eve::unalign(data_aligned()); }
+
+    void swap(soa_storage& other) noexcept
+    {
+      storage_.swap(other.storage_);
+      std::swap(indexes_,other.indexes_);
+      std::swap(size_,other.size_);
+    }
+
+    void build(std::size_t n, Type v = {})
+    {
+      using alloc_t = std::allocator_traits<Allocator>;
+      auto sub_span = eve::align(&storage_[0], eve::over{64});
+
+      kumi::for_each( [&]<typename T>(T*& s, auto m)
+                      {
+                        s = reinterpret_cast<T*>(sub_span);
+                        for(std::size_t i=0;i<n;++i) alloc_t::construct(*this, s+i,m);
+                        sub_span += realign<T>(n);
+                      }
+                    , indexes_, v
+                    );
+
+      capacity_ = capacity_impl(n);
+    }
+
+    void reindex(std::size_t n)
+    {
+      auto sub_span = eve::align(&storage_[0], eve::over{64});
+
+      kumi::for_each( [&]<typename T>(T*& s)
+                      {
+                        s = reinterpret_cast<T*>(sub_span);
+                        sub_span += realign<T>(n);
+                      }
+                    , indexes_
+                    );
+      capacity_ = capacity_impl(n);
+    }
+
+    //==============================================================================================
+    // Static helpers
+    //==============================================================================================
+    template<typename Content>
+    static void piecewise_copy(Content& dst, Content const& src, std::size_t n)
+    {
+      kumi::for_each( [&]<typename T>(T* d, T const* s) { std::memcpy(d,s,sizeof(T)*n); }
+                    , dst, src
+                    );
+    }
+
+    template<typename T> static EVE_FORCEINLINE auto as_aligned_pointer(T* ptr)
+    {
+      return eve::as_aligned(ptr, eve::detail::cache_line_cardinal<T>{});
+    }
+
+    template<typename T>
+    static  auto realign(std::size_t n) { return eve::align(sizeof(T)*n, eve::over{64}); }
+
+    static auto capacity_impl(std::size_t n) noexcept
+    {
+      // Capacity is the smallest gap possible, ie the one in the largest type
+      constexpr std::size_t largest = max_scalar_size_v<Type>;
+      return eve::align(largest*n, eve::over{64})/largest;
+    }
+
+    static auto sizing(std::size_t n) noexcept
+    {
+      return  n ? 64 + kumi::fold_left( [n]<typename S>(auto a, S) { return a + realign<S>(n); }
+                                  , flat_t{}
+                                  , 0ULL
+                                  )
+                : n;
+    }
```

Comment:
I do not get these 2 functions.
Like - capacity in what? Elements? bytes?

Reply:
elements always. It si used to compute the actual capacity() of the soa_vector

Reply:
Renamed both to:

*  `available_elements` for the capacity_impl one
* `byte_size` for the sizing one

Reply:
The problem is the code, I find the code difficult to understand.

constexpr std::size_t largest = max_scalar_size_v<Type>;
return eve::align(largest*n, eve::over{64})/largest;

This makes no sense to me.